crates/circuits/sha256-air/src/sha256/trace.rs

```diff
@@ -0,0 +1,405 @@
+                        bitwise_lookup_chip.request_range(carry_a, carry_e);
+                    }
+
+                    // update working variables
+                    work_vars[7] = work_vars[6];
+                    work_vars[6] = work_vars[5];
+                    work_vars[5] = work_vars[4];
+                    work_vars[4] = e;
+                    work_vars[3] = work_vars[2];
+                    work_vars[2] = work_vars[1];
+                    work_vars[1] = work_vars[0];
+                    work_vars[0] = a;
+                }
+            }
+            // computing the final hash
+            else {
+                let cols: &mut Sha256DigestCols<F> = row[..SHA256_DIGEST_WIDTH].borrow_mut();
+                cols.flags.is_round_row = F::ZERO;
+                cols.flags.is_first_4_rows = F::ZERO;
+                cols.flags.is_digest_row = F::ONE;
+                cols.flags.is_last_block = F::from_bool(is_last_block);
+                cols.flags.row_idx =
+                    get_flag_pt_array(&self.row_idx_encoder, 16).map(F::from_canonical_u32);
+                cols.flags.global_block_idx = F::from_canonical_u32(global_block_idx);
+
+                cols.flags.local_block_idx = F::from_canonical_u32(local_block_idx);
+                final_hash = array::from_fn(|i| work_vars[i].wrapping_add(prev_hash[i]));
+                cols.final_hash = array::from_fn(|i| {
+                    u32_into_limbs::<SHA256_WORD_U8S>(final_hash[i]).map(F::from_canonical_u32)
+                });
+                cols.prev_hash = prev_hash
+                    .map(|f| u32_into_limbs::<SHA256_WORD_U16S>(f).map(F::from_canonical_u32));
+                let hash = if is_last_block {
+                    SHA256_H.map(u32_into_limbs::<SHA256_WORD_BITS>)
+                } else {
+                    cols.final_hash
+                        .map(|f| limbs_into_u32(f.map(|x| x.as_canonical_u32())))
+                        .map(u32_into_limbs::<SHA256_WORD_BITS>)
+                }
+                .map(|x| x.map(F::from_canonical_u32));
+
+                for i in 0..SHA256_ROUNDS_PER_ROW {
+                    cols.hash.a[i] = hash[SHA256_ROUNDS_PER_ROW - i - 1];
+                    cols.hash.e[i] = hash[SHA256_ROUNDS_PER_ROW - i + 3];
+                }
+            }
+        }
+        let (first16, last1) = output.split_at_mut(16);
+        let local_cols: &Sha256RoundCols<F> = first16[15].as_slice().borrow();
+        let next_cols: &mut Sha256RoundCols<F> = last1[0].as_mut_slice().borrow_mut();
+        Self::generate_carry_ae(local_cols, next_cols, 16);
+        (output, final_hash)
+    }
+
+    /// Puts the correct carrys in the `next_row`, the resulting carrys can be out of bound
+    /// Here, row_idx is the index of the `next_row` inside the block: 0..16 for the first 16 rows and
+    /// some value outside of the 0..16 range for other rows (possibly invalid)
+    fn generate_carry_ae<F: PrimeField32>(
```

Comment:
it feels like this has a lot of duplication with the work vars calculation and should ideally be inlined there -- is there a blocker?